### PR TITLE
lesson04br

### PR DESCRIPTION
**Pull request by vitalyryzhkov**

### CODE REVIEW

File: src/lesson04/List.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package lesson04;
+
+public interface List<T> {
+    T get(int i);
```
Comment (Anderwerty) [inline src/lesson04/List.java:4]: Better not `i `but` index`. The same for the rest code.

File: src/lesson04/List.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+    void add(T t);
```
Comment (Anderwerty) [inline src/lesson04/List.java:5]: And not t but item or element

File: src/lesson04/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void remove(int i) {
+        if (array.length - 1 - i >= 0) {
+            System.arraycopy(array, i + 1, array, i, array.length - 1 - i);
+        }
+        array[array.length - 1] = null;
+    }
+
+    @Override
+    public int size() {
```
Comment (Anderwerty) [inline src/lesson04/MyArrayList.java:55]: It is not a size of the list it is capacity.  You should add one more field. For example size. Size= number of elements you put into the list. 
Please read https://habr.com/ru/post/128269/ and reimplement methods

File: src/lesson04/MyLinkedList.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        Node<T> currentNode = first;
+        if (currentNode == null) {
+            Node<T> last = this.last;
+            Node<T> newNode = new Node<>(last, t, null);
+            this.last = newNode;
+            if (last == null)
+                currentNode = newNode;
+            else
+                last.next = newNode;
+            size++;
```
Comment (Anderwerty) [inline src/lesson04/MyLinkedList.java]: You can see size++ in if and else blocks. So it possible to remove line 30 and  line 35. And put size++ after if-else block.

File: src/lesson04/MyLinkedList.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    @Override
+    public void clear() {
+        for (Node<T> x = first; x != null; ) {
+            Node<T> next = x.next;
+            x.item = null;
+            x.next = null;
+            x.prev = null;
+            x = next;
+        }
+        first = last = null;
```
Comment (Anderwerty) [inline src/lesson04/MyLinkedList.java]: It is not a java style.  Should be 
first=null;
last=null;

File: src/lesson04/MyLinkedList.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        size = 0;
+    }
+
```
Comment (Anderwerty) [inline src/lesson04/MyLinkedList.java:77]: Remove redundant blank line.

File: src/lesson04/MyLinkedList.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        } else {
+            Node<T> elem = last;
+            for (int i = size - 1; i > index; i--) {
+                elem = elem.prev;
+            }
+            return elem;
+        }
+    }
+
+    static class Node<E> {
```
Comment (Anderwerty) [inline src/lesson04/MyLinkedList.java]: Why no private?

File: src/lesson04/MyStack.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
+package lesson04;
+
+public class MyStack<T> {
```
Comment (Anderwerty) [inline src/lesson04/MyStack.java]: Create interface Stack and implement its methods in the class MyStack

File: src/lesson04/MyStack.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    private int size;
+    private Object[] array;
+    private int space;
```
Comment (Anderwerty) [inline src/lesson04/MyStack.java:6]: What is space? Explain, please.

**Review by Anderwerty — CHANGES_REQUESTED** [review]
Resolve comments

File: src/lesson04/MyStack.java (hunk 9 above)
Comment (vitalyryzhkov) [reply src/lesson04/MyStack.java:6]: не смогу объяснить на английском, но это что-то вроде "второго размера" - "объем", скорее всего не очень удачное название, да? лучше capacity?

**Review by Anderwerty — APPROVED** [review]

**Pull request merged**